flax/cursor.py

```diff
@@ -0,0 +1,341 @@
+      obj = self.obj.copy(changes)  # type: ignore
+    elif isinstance(self.obj, (dict, list)):
+      obj = self.obj.copy()  # type: ignore
+      for key, value in changes.items():
+        obj[key] = value
+    elif isinstance(self.obj, tuple):
+      obj = list(self.obj)  # type: ignore
+      for key, value in changes.items():
+        obj[key] = value
+      obj = tuple(obj)  # type: ignore
+    elif dataclasses.is_dataclass(self.obj):
+      obj = dataclasses.replace(self.obj, **changes)  # type: ignore
+    else:
+      obj = self.obj  # type: ignore
+
+      # NOTE: There is a way to try to do a general replace for pytrees, but it requires
+      # the key of `changes` to store the type of access (getattr, getitem, etc.)
+      # in order to access those value from the original object and try to replace them
+      # with the new value. For simplicity, this is not implemented for now.
+      # ----------------------
+      # changed_values = tuple(changes.values())
+      # result = flatten_until_found(self.obj, changed_values)
+
+      # if result is None:
+      #   raise ValueError('Cannot find object in parent')
+
+      # leaves, treedef = result
+      # leaves = [leaf if leaf is not self.obj else value for leaf in leaves]
+      # obj = jax.tree_util.tree_unflatten(treedef, leaves)
+
+    return obj  # type: ignore
+
+  def apply_update(
+      self,
+      update_fn: Callable[[Sequence[Union[str, int]], Any], Tuple[bool, Any]],
+      _path: Tuple[Union[str, int], ...] = tuple(),
+  ) -> 'Cursor[A]':
+    """Traverse the Cursor object and apply conditional changes recursively via an ``update_fn``.
+    The ``update_fn`` has a function signature of ``(Sequence[Union[str, int]], Any) -> Tuple(bool, Any)``:
+
+    - The input arguments are the current key path and value at that current key path
+    - The output values are a bool denoting whether to set the Cursor object at the current
+      key path to the new value, and the new value itself
+
+    To generate a copy of the Cursor object with the accumulated changes, call the `.build` method.
+
+    NOTE: If the ``update_fn`` returns True, this function will not recurse any further down that branch
+    to apply changes. For example, if we intend to replace an attribute that points to a dictionary
+    with an int, we don't need to look for further changes inside the dictionary, since the dictionary
+    will be replaced anyways.
+
+    Example::
+
+      import flax.linen as nn
+      from flax.cursor import cursor
+      import jax
+      import jax.numpy as jnp
+
+      class Model(nn.Module):
+        @nn.compact
+        def __call__(self, x):
+          x = nn.Dense(3)(x)
+          x = nn.relu(x)
+          x = nn.Dense(3)(x)
+          x = nn.relu(x)
+          x = nn.Dense(3)(x)
+          x = nn.relu(x)
+          return x
+
+      params = Model().init(jax.random.PRNGKey(0), jnp.empty((1, 2)))['params']
+
+      def update_fn(path, value):
```

Comment:
why return both boolean and possibly updated value?  what's the optimization here?

Reply:
My bad, if we use `id` to detect changes it should work.